rosbag2_cpp/src/rosbag2_cpp/cache/message_cache_circular_buffer.cpp

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <list>
+
+#include "rosbag2_cpp/cache/message_cache_circular_buffer.hpp"
+
+namespace rosbag2_cpp
+{
+namespace cache
+{
+/**
+* This class implements a circular buffer message cache. Since the buffer
+* size is limited by total byte size of the storage messages rather than
+* a fix number of messages, a linked list is used instead of a vector since
+* older messages can always be dropped from the front and new messages added
+* to the end. This buffer may consume more byte than max_cache_size, however
+* it will only exceed the limit by the excess data in the most recent message.
+*/
```

Comment:
It does make sense to move description to the class in relevant header `hpp` file, where all other interface descriptions and doxygen comments.